File: src/lib/fileConverter.ts
// src/lib/fileConverter.ts
// Stage 1: File Conversion — DWG/PDF → PNG via GroupDocs Cloud API
//
// GroupDocs Conversion Cloud handles the heavy CAD/PDF rendering.
// Images (JPG/PNG) pass through without conversion.

import * as groupdocs_conversion_cloud from 'groupdocs-conversion-cloud';

const PASSTHROUGH_MIMES = ['image/png', 'image/jpeg', 'image/jpg'];

const CONVERTIBLE_MIMES: Record<string, string> = {
    'application/pdf': 'pdf',
    'image/vnd.dwg': 'dwg',
    'application/acad': 'dwg',
    'application/x-dwg': 'dwg',
    'application/dxf': 'dxf',
    'application/x-dxf': 'dxf',
};

// Also check file extension as a fallback (browsers may not set correct MIME for DWG)
const EXTENSION_MAP: Record<string, string> = {
    '.dwg': 'dwg',
    '.dxf': 'dxf',
    '.pdf': 'pdf',
    '.png': 'png',
    '.jpg': 'jpg',
    '.jpeg': 'jpg',
};

export interface ConversionResult {
    imageBuffers: Buffer[];
    originalFormat: string;
    wasConverted: boolean;
    error?: string;
}

function getFormatFromExtension(fileName: string): string | null {
    const ext = fileName.toLowerCase().slice(fileName.lastIndexOf('.'));
    return EXTENSION_MAP[ext] ?? null;
}

export async function convertToPng(
    fileBuffer: Buffer,
    mimeType: string,
    fileName: string
): Promise<ConversionResult> {
    // If already an image, pass through
    if (PASSTHROUGH_MIMES.includes(mimeType)) {
        return { imageBuffers: [fileBuffer], originalFormat: mimeType, wasConverted: false };
    }

    // Check MIME type first, then fallback to extension
    let sourceFormat = CONVERTIBLE_MIMES[mimeType];
    if (!sourceFormat) {
        sourceFormat = getFormatFromExtension(fileName) ?? '';
    }

    if (!sourceFormat || ['png', 'jpg', 'jpeg'].includes(sourceFormat)) {
        // It's an image by extension but didn't match MIME — pass through
        if (['png', 'jpg', 'jpeg'].includes(sourceFormat || '')) {
            return { imageBuffers: [fileBuffer], originalFormat: sourceFormat || mimeType, wasConverted: false };
        }
        return {
            imageBuffers: [],
            originalFormat: mimeType,
            wasConverted: false,
            error: `Unsupported file type: ${mimeType} (${fileName}). Accepted: PDF, DWG, DXF, JPG, PNG.`,
        };
    }

    // ── Local PDF Conversion via pdfjs-dist ──
    if (sourceFormat === 'pdf') {
        try {
            console.log('[FileConverter] Using local pdfjs-dist for PDF conversion');
            
            // pdfjs-dist requires a specific import for Node.js environments
            const pdfjsLib = require('pdfjs-dist/legacy/build/pdf.js');
            const Canvas = require('canvas');

            const loadingTask = pdfjsLib.getDocument({ data: new Uint8Array(fileBuffer) });
            const pdfDocument = await loadingTask.promise;
            
            const imageBuffers: Buffer[] = [];
            const numPages = Math.min(pdfDocument.numPages, 3); // Max 3 pages for floor plans

            for (let pageNum = 1; pageNum <= numPages; pageNum++) {
                const page = await pdfDocument.getPage(pageNum);
                
                // Use a scale that provides good resolution (e.g. 150-200 DPI equivalent)
                const scale = 2.0; 
                const viewport = page.getViewport({ scale });

                const canvas = Canvas.createCanvas(viewport.width, viewport.height);
                const context = canvas.getContext('2d');

                const renderContext = {
                    canvasContext: context as any, // Cast needed due to DOM types mismatch with Node canvas
                    viewport: viewport,
                };

                await page.render(renderContext).promise;
                
                // Convert Node canvas to PNG buffer
                const pngBuffer = canvas.toBuffer('image/png', { compressionLevel: 3 });
                imageBuffers.push(pngBuffer);
            }

            if (imageBuffers.length > 0) {
                return {
                    imageBuffers,
                    originalFormat: sourceFormat,
                    wasConverted: true,
                };
            } else {
                throw new Error("Local PDF conversion produced no output.");
            }

        } catch (pdfErr: any) {
            console.error('[pdfjs Error]:', pdfErr);
            return {
                imageBuffers: [],
                originalFormat: sourceFormat,
                wasConverted: false,
                error: `Local PDF conversion failed: ${pdfErr.message || 'Unknown PDF processing error'}.`,
            };
        }
    }

    // ── GroupDocs Fallback for DWG/DXF ──
    const clientId = process.env.GROUPDOCS_CLIENT_ID;
    const clientSecret = process.env.GROUPDOCS_CLIENT_SECRET;

    if (!clientId || !clientSecret) {
        return {
            imageBuffers: [],
            originalFormat: sourceFormat,
            wasConverted: false,
            error: 'GroupDocs API credentials not configured. Setup GROUPDOCS_CLIENT_ID and GROUPDOCS_CLIENT_SECRET in .env.local to convert CAD files.',
        };
    }

    try {
        console.log('[FileConverter] Using GroupDocs for CAD conversion');
        // Initialize GroupDocs API
        const config = new groupdocs_conversion_cloud.Configuration(clientId, clientSecret);
        const fileApi = groupdocs_conversion_cloud.FileApi.fromConfig(config);
        const convertApi = groupdocs_conversion_cloud.ConvertApi.fromConfig(config);

        // Upload file to GroupDocs cloud storage
        const uploadPath = `firerulx/uploads/${Date.now()}_${fileName}`;
        const uploadRequest = new groupdocs_conversion_cloud.UploadFileRequest(uploadPath, fileBuffer);
        await fileApi.uploadFile(uploadRequest);

        // Convert to PNG
        const settings = new groupdocs_conversion_cloud.ConvertSettings();
        settings.filePath = uploadPath;
        settings.format = 'png';
        settings.outputPath = `firerulx/converted/${Date.now()}_output.png`;

        const convertRequest = new groupdocs_conversion_cloud.ConvertDocumentRequest(settings);
        const result = await convertApi.convertDocument(convertRequest);

        if (!result || result.length === 0) {
            return {
                imageBuffers: [],
                originalFormat: sourceFormat,
                wasConverted: false,
                error: 'GroupDocs conversion returned no output.',
            };
        }

        // Download all converted PNG pages
        const imageBuffers: Buffer[] = [];
        for (const res of result) {
            if (res.path) {
                const downloadRequest = new groupdocs_conversion_cloud.DownloadFileRequest(res.path);
                const downloadResult = await fileApi.downloadFile(downloadRequest);
                imageBuffers.push(Buffer.from(downloadResult));
            }
        }

        // Clean up uploaded file (best effort)
        try {
            const deleteRequest = new groupdocs_conversion_cloud.DeleteFileRequest(uploadPath);
            await fileApi.deleteFile(deleteRequest);
        } catch {
            // Non-critical: cleanup failure is OK
        }

        return {
            imageBuffers,
            originalFormat: sourceFormat,
            wasConverted: true,
        };
    } catch (err: any) {
        console.error('[GroupDocs Conversion Error]:', err);
        
        let message = 'Unknown conversion error';
        if (err instanceof Error) {
            message = err.message;
        } else if (typeof err === 'object' && err !== null) {
            if (err.message) message = err.message;
            else if (err.error && err.error.message) message = err.error.message;
            else {
                try { message = JSON.stringify(err); } catch { message = String(err); }
            }
        } else {
            message = String(err);
        }

        return {
            imageBuffers: [],
            originalFormat: sourceFormat,
            wasConverted: false,
            error: `CAD conversion failed (GroupDocs): ${message}`,
        };
    }
}

File: backend/file_converter.py
# backend/file_converter.py
# Stage 1: File Conversion — PDF → PNG via PyMuPDF
#
# Images (JPG/PNG) pass through without conversion.
# PDFs are rendered to PNG using PyMuPDF (fitz).
# DWG/DXF are currently unsupported.

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Optional

import fitz  # PyMuPDF

PASSTHROUGH_MIMES = {"image/png", "image/jpeg", "image/jpg"}

CONVERTIBLE_MIMES: dict[str, str] = {
    "application/pdf": "pdf",
    "image/vnd.dwg": "dwg",
    "application/acad": "dwg",
    "application/x-dwg": "dwg",
    "application/dxf": "dxf",
    "application/x-dxf": "dxf",
}

EXTENSION_MAP: dict[str, str] = {
    ".dwg": "dwg",
    ".dxf": "dxf",
    ".pdf": "pdf",
    ".png": "png",
    ".jpg": "jpg",
    ".jpeg": "jpg",
}


@dataclass
class ConversionResult:
    image_buffers: list[bytes] = field(default_factory=list)
    original_format: str = ""
    was_converted: bool = False
    error: Optional[str] = None


def _get_format_from_extension(file_name: str) -> Optional[str]:
    _, ext = os.path.splitext(file_name.lower())
    return EXTENSION_MAP.get(ext)


def convert_to_png(
    file_buffer: bytes, mime_type: str, file_name: str
) -> ConversionResult:
    """Convert file to PNG image(s). Images pass through; PDFs rendered via PyMuPDF."""

    # If already an image, pass through
    if mime_type in PASSTHROUGH_MIMES:
        return ConversionResult(
            image_buffers=[file_buffer],
            original_format=mime_type,
            was_converted=False,
        )

    # Check MIME type first, then fallback to extension
    source_format = CONVERTIBLE_MIMES.get(mime_type)
    if not source_format:
        source_format = _get_format_from_extension(file_name) or ""

    if not source_format or source_format in ("png", "jpg", "jpeg"):
        if source_format in ("png", "jpg", "jpeg"):
            return ConversionResult(
                image_buffers=[file_buffer],
                original_format=source_format or mime_type,
                was_converted=False,
            )
        return ConversionResult(
            original_format=mime_type,
            was_converted=False,
            error=f"Unsupported file type: {mime_type} ({file_name}). Accepted: PDF, JPG, PNG.",
        )

    # ── Local PDF Conversion via PyMuPDF ──
    if source_format == "pdf":
        try:
            doc = fitz.open(stream=file_buffer, filetype="pdf")
            image_buffers: list[bytes] = []
            num_pages = min(doc.page_count, 3)  # Max 3 pages

            for page_num in range(num_pages):
                page = doc.load_page(page_num)
                # Prevent OOM by scaling dynamically. Max dimension ~1536px.
                rect = page.rect
                zoom = min(1536.0 / max(rect.width, 1.0), 1536.0 / max(rect.height, 1.0), 2.0)
                mat = fitz.Matrix(zoom, zoom)
                pix = page.get_pixmap(matrix=mat, alpha=False, colorspace=fitz.csRGB)
                png_bytes = pix.tobytes("png")
                image_buffers.append(png_bytes)

            doc.close()

            if image_buffers:
                return ConversionResult(
                    image_buffers=image_buffers,
                    original_format=source_format,
                    was_converted=True,
                )
            else:
                raise RuntimeError("Local PDF conversion produced no output.")

        except Exception as e:
            return ConversionResult(
                original_format=source_format,
                was_converted=False,
                error=f"Local PDF conversion failed: {e}",
            )

    # ── DWG/DXF — unsupported ──
    return ConversionResult(
        original_format=source_format,
        was_converted=False,
        error=f"CAD file format ({source_format.upper()}) is not supported in the current backend. Please convert to PDF or PNG first.",
    )
